mautic/mautic
# API contact merge fix

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | N
| Automated tests included? | in API Library
| Related user documentation PR URL | /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | Y

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR is solving 2 issues:
1. The API Library edit PUT test was failing on trying to merge a contact entity with `NULL`. The problem was that it did not check if the specified contact exist before it tried to merge. The solution I suggest is to let the edit method check if the entity exist first and merge duplicated contacts later in `preSaveEntity` method.

2. Mautic was creating duplicate contacts when saving contacts with invalid email address.

Try to send this payload:
```
[
  {
    "email": "email1@email.com",
    "firstname": "BatchUpdate"
  },
  {
    "email": "email2@email.com,",
    "firstname": "BatchUpdate2"
  },
  {
    "email": "email3@email.com",
    "firstname": "BatchUpdate3"
  },
  {
    "email": "email2@email.com,",
    "firstname": "BatchUpdate2"
  }
]
```
to `POST api/contacts/batch/new`

Notice the email2... is an invalid email address with comma at the end. Mautic will check whether the contact with such email address exists and in this case it will never exist. Then it cleans the values and saves the contacts with the fixed email address.

The solution is to clean values before searching for duplicates.

While in it I noticed there is quite lot of code duplicated so I adjusted also other methods to be more DRY.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Checkout this [API Library PR](https://github.com/mautic/api-library/pull/142) because it contains some unit test fixes.
2. Run unit tests specific for contacts to save your time `vendor/bin/phpunit --filter ContactsTest --stop-on-failure`
3. Notice the failure.
4. Run the API request which is causing duplicates (above) and check it creates duplicates.

#### Steps to test this PR:
1. Checkout this PR
2. Run the API Library tests again - they should pass.
3. Run the problematic API request again - no new duplicates will be created.

#### List deprecations along with the new alternative:
1. `LeadApiController::getExistingLead()` does almost the same as `LeadModel:: checkForDuplicateContact()` so I changed the code to use the method from model and deprecated the duplicated method.
